improve axis3 handling

bwinkel opened the pull request:
Remove the need to patch the WCS header (to make it 3D), even if one is
only interested in a 2D map. Can now specify a keyword-arg ('naxis3')
for that. Furthermore, the grid-method automatically adds a 2nd dimension
to the signal and weight arrays, if necessary.

Fix #10

bwinkel committed:
improve axis3 handling

Remove the need to patch the WCS header to make it 3D, even if one is
only interested in a 2D map. Can now specify a keyword-arg ('naxis3')
for that. Furthermore, the grid-method automatically adds a 2nd dimension
to the signal and weight arrays, if necessary.

DanielLenz commented:
Thanks for working on this, will check it out later this week!


> On Feb 24, 2019, at 1:34 PM, Benjamin Winkel <notifications@github.com> wrote:
> 
> @bwinkel <https://github.com/bwinkel> requested your review on: #12 <https://github.com/bwinkel/cygrid/pull/12> improve axis3 handling.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/bwinkel/cygrid/pull/12#event-2160348062>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AHLvZ_cxRuRsd94IzVqePTusIMTSsoR1ks5vQwVmgaJpZM4bOyOm>.
>

DanielLenz commented on cygrid/cygrid.pyx:
```diff
@@ -900,7 +917,7 @@ cdef class SlGrid(Cygrid):
     sl_lons, sl_lats : numpy.ndarray (float64), 1D
         Coordinates of sight lines to grid onto.
-    naxes3 : int
+    naxes3 : int (Default: 1)
```
Could you be more verbose about the default being extracted from the header, and only set to 1 if there is no such entry?

DanielLenz commented on cygrid/cygrid.pyx:
```diff
@@ -976,7 +993,7 @@ cdef class SlGrid(Cygrid):
             np.ndarray[double, ndim=1] sl_lons,
             np.ndarray[double, ndim=1] sl_lats,
-            long naxes3,
+            long naxes3=1,
```
I'm a bit confused by the naxes3 keyword. Shouldn't this be naxis3 or naxes? And why exactly is the default value 1? Or is this meant to be boolean (ie data is cube if naxes3=True, map if naxes3=False)?

DanielLenz commented on cygrid/cygrid.pyx:
```diff
@@ -1042,7 +1059,7 @@ cdef class SlGrid(Cygrid):
             The gridded spectral data for each of the input positions.
         '''
-        return self.datacube / self.weightscube
+        return self.datacube[:, 0] / self.weightscube[:, 0]
```
Same as comment above; I don't immediately understand the slicing here. Does datacube[:, 0] extract a list of spectra?

DanielLenz commented on cygrid/cygrid.pyx:
```diff
@@ -1053,7 +1070,7 @@ cdef class SlGrid(Cygrid):
             The gridded spectral weights for each of the input positions.
         '''
-        return self.weightscube
+        return self.weightscube[:, 0]
```
Same as above

DanielLenz commented on cygrid/tests/test_cygrid.py:
```diff
@@ -58,26 +62,132 @@ def setup(self):
+        mygridder = cygrid.WcsGrid(header, **kwargs)
+        mygridder.set_kernel(*self.kernel_args)
+
+        mygridder.grid(self.xcoords, self.ycoords, signal[:, np.newaxis])
```
Wasn't the purpose of this patch to get rid of the degenerate axis if one wants to grid maps? Sorry if I'm misunderstanding this!

DanielLenz commented on cygrid/tests/test_cygrid.py:
```diff
@@ -58,26 +62,132 @@ def setup(self):
+
+        header = self.header.copy()
+        header['NAXIS'] = 3
+        header['NAXIS3'] = 1
```
Okay, after looking at the test cases, I think I understand things much more clearly.
I'm sure you've already looked into this, but is there a way to automatically set naxis3 (why is it naxis3 here, and nax**e**s3 above?), based on the shape of the input data?

DanielLenz commented on cygrid/tests/test_cygrid.py:
```diff
@@ -58,26 +62,132 @@ def setup(self):
+
+    def test_gridding2d_naxis3_auto(self):
+        '''
+        Automatic handling of naxis3 (aka set to One, if not present)
```
Okay, fully understood it now! Took me a while, haven't had enough coffee yet...

DanielLenz reviewed (approved):
Great work, thanks for implementing this! Please read all my comments before addressing anything - I was confused at first, but things cleared up as I read through the changes.

One last item is the naming of naxes3 vs naxis3 (I think the former shouldn't be used), could you clear that up?

Also thanks for cleaning up the examples and removing all mentions of the artificial third axis!

Happy to skype this week as well, if you want to discuss any of this. It's been a while :)

bwinkel commented:
Thanks for the thorough review. You raised a good point. For historical reasons, `cygrid` prepares the datacube arrays in the constructor (which calls the `_prepare` method). This has two advantages: (1) we can check if the raw data is consistent with the desired target datacube shape, and (2) `cygrid` fails early, if not enough memory is present to handle the data- and weightcubes. However, none of these two points seems utterly important to me, at least not more important than a user interface that is as simple as possible. 

I _think_, it would be possible to defer the datacube allocation to the `grid`-method call and derive the z-axis size from the raw data array. One could even store the shape of the raw data array and have the `get_datacube` and `get_weights` return a shape, which is consistent with that, e.g., if the raw data array is 2D, the returned datacube should be 3D, if raw data is just 1D, one probably wants a 2D datacube (aka map).

I will think about a possible implementation and get back to you. If you'd rather prefer the current API, please let me know asap.

By-the-way, I just realized that `cygrid` - when provided with a datacube in the constructor - has a high chance of creating a new datacube instance, due to the possible alterations in the `grid` method (e.g., to change bit-length, 32 or 64, depending on raw data). This is something, which one should handle correctly and add some tests.

bwinkel committed:
bugfix: tests used an ill-defined kernel width

bwinkel committed:
add tests that check if datacube is re-instantiated when provided by user

currently, this is the case (but needs to be fixed)

bwinkel committed:
prepare new handling of naxis3

split `_prepare` method into a spatial-coord and a datacube part
the latter is now called at the first call of the `grid` method

the idea is to take input signal shape and adjust datacube accordingly;
then we would need to handle the naxis3 any more

Commit:
new internal data array handling

now works with every input signal shape; internal datacube is properly
casted to a 3D object and in the get_datacube method everything
is shaped to what the user indicated with the input signal shape

bwinkel commented:
Well, that was a substantial amount of work, but I managed to completely remove all manual naxis3 handling from the user side. One can now provide any shape for the input raw data signal, this is magically reshaped to a 3D array for internal use and the `get_datacube` methods will again apply a reshape to make the output consistent with the input signal.

For example, if signal has shape `(200, 10)` and the map is `(50, 50)`, the returned datacube will be `(10, 50, 50)`. This works with any dimensions (aka Stokes parameters are now working). It was quite challenging to get it working with both `WcsGrid` and `SlGrid` at the same time...

I need to update all the docs, still. Also, unfortunately, this change is not fully backwards compatible. If one still has the `grid(..., signal[:, np.newaxis])` in the code, `cygrid` will now produce an extra dimension in the output.

Commit:
bugfix: fix byteorder of input data/weights if not native

apparently, astroquery.skyview can deliver non-native byteorders

Commit:
remove duplicate definition of MAX_Y from cygrid.pyx

Commit:
update docs

Commit:
update notebooks

bwinkel commented:
Can I ask you to have another look, @DanielLenz? I'm sorry to steal your time (after all, it was your idea 😁).

DanielLenz commented:
Wow, thanks for working on all of this so promptly! I'll give it a careful review when I can, hopefully within the next few days.

I think it's fair to break the API for 2.0, these are really important changes that improve the handling quite a bit.
